### ISSUE TITLE
Better use of STUDIO_ enums and deprecation of halo.RenderedEntity

### ISSUE THREAD

**Issue by NeatNit** [Facepunch/garrysmod-requests#948]
The [studio enums](http://wiki.garrysmod.com/page/Enums/STUDIO) are currently only used when [ENT:Draw](http://wiki.garrysmod.com/page/ENTITY/Draw) is called by the engine. For example, when an entity is rendered normally, Draw is called with value 1 (STUDIO_RENDER) and when it's rendered for a projected texture, it's called with value 1073741825 (STUDIO_RENDER and STUDIO_SHADOWDEPTHTEXTURE). Presumably it's used similarly in other places, but I didn't look into it besides those 2 scenarios.

[halo.RenderedEntity](http://wiki.garrysmod.com/page/halo/RenderedEntity) serves a similar function in a crappy way. In vanilla gmod it's only used in [gmod_thruster](https://github.com/garrynewman/garrysmod/blob/master/garrysmod/gamemodes/sandbox/entities/entities/gmod_thruster.lua#L93), and you're expected to check with the halo library whether your entity is only being rendered for a stencil. This technically works, but no user-made libraries can make use of the same "tell-entities-not-to-draw-non-stencil-stuff" ability unless they detour halo.RenderedEntity. It's bullshit.

I propose:
1. Allow the user to specify STUDIO flags when calling [Entity:DrawModel](http://wiki.garrysmod.com/page/Entity/DrawModel)
2. Add a flag to specify stencil drawing, e.g. STUDIO_STENCIL
3. Make sure that [ENT:DrawTranslucent](http://wiki.garrysmod.com/page/ENTITY/DrawTranslucent) and [ENT:RenderOverride](http://wiki.garrysmod.com/page/ENTITY/RenderOverride) also get the STUDIO enums as an argument, if they don't already
4. Deprecate halo.RenderedEntity and make it use the new STUDIO_STENCIL flag when rendering entities

**Comment by Jvs34** [comment]
I wanted to make a request about this for a long time.

I agree on the ent:DrawModel change, along with adding the flags on Pre/PostPlayerDraw hooks, so you can pass down the flags to the entity you might manually draw ( in my case I've got a jetpack with a ENT:DrawTranslucent  that won't get called during player draw as it counts as a normal opaque draw ).

**Comment by NeatNit** [comment]
Yeah, all relevant render hooks should get it. Most prominently PreDraw* and PostDraw*.

**Comment by robotboy655** [comment]
This was partially addressed,

Added new optional argument to Entity.DrawModel() - flags (STUDIO_)
PrePlayerDraw/PostPlayerDraw now have 2nd argument - flags (STUDIO_)
DrawWorldModelTranslucent/DrawWorldModel now have 2nd argument - flags (STUDIO_)
RenderOverride now also has 2nd argument - flags (STUDIO_)
